[C++ API] Add functional overloads for fold, linear, loss, normalization, padding

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29360 Add functional overloads for fold, linear, loss, normalization, padding**
* #29359 Add functional overloads for pixelshuffle, pooling, upsampling, vision
* #29358 Add functional overloads for activation, batchnorm, distance, embedding

This PR adds functional overloads that take the full set of arguments (instead of just Options) for the following functionals:
- fold
- linear
- loss
- normalization
- padding

These new functionals lives in the `torch::nn::functional::detail` namespace and they are only meant to be called from the module forward methods (i.e. they are not public API). This is in preparation for the future change where we make module Options and functional Options two different classes, because if the module forward method has to construct a new functional Options object every time it runs it will be pretty silly and bad performance.

Differential Revision: [D18376975](https://our.internmc.facebook.com/intern/diff/D18376975)

Commit:
Add functional overloads for fold, linear, loss, normalization, padding

Commit:
Update on "Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

Commit:
Update on "Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

Commit:
Update on "Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

kostmo commented:
## CircleCI build failures summary
As of commit 5f2783f:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=5f2783f4ff18b153914114b6a76995fd9d938a2c).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "[C++ API] Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

Commit:
Update on "[C++ API] Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

Commit:
Update on "[C++ API] Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

Commit:
Update on "[C++ API] Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

Commit:
Update on "[C++ API] Add functional overloads for fold, linear, loss, normalization, padding"

Add functional overloads for fold, linear, loss, normalization, padding

gh-metadata: pytorch pytorch 29360 gh/yf225/39/head

yf225 commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -68,16 +91,28 @@ inline Tensor mse_loss(
     const Tensor& target,
-    const BCELossOptions& options = {}) {
-  auto reduction_enum = enumtype::reduction_get_enum(options.reduction());
+    Tensor weight,
```
thanks a lot for the catch! I changed it to pass by const reference

yf225 commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -91,67 +126,116 @@ inline Tensor binary_cross_entropy(
       "!= input nelement (", input.numel(), ")");
   }
 
-  auto weight = options.weight();
```
right it doesn't, although I changed `weight` to pass by const reference just to be safe

yf225 commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -91,67 +126,116 @@ inline Tensor binary_cross_entropy(
-    TORCH_CHECK(options.weight().dim() == 1, "weight must be one-dimensional");
+    int64_t p,
+    double margin,
+    Tensor weight,
```
thanks a lot for the catch! I changed it to pass by const reference

yf225 commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -162,105 +246,226 @@ inline Tensor smooth_l1_loss(
     const Tensor& input,
     const Tensor& target,
-    const MultiLabelSoftMarginLossOptions& options = {}) {
+    Tensor weight,
```
thanks a lot for the catch! I changed it to pass by const reference
